### PR TITLE
[Feature] add yolox ncnn

### PR DESCRIPTION
**Pull request by hanrui1sensetime**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily receiving feedbacks. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Support YOLOX model for ncnn.
The evaluation for coco dataset is 32.0AP, the same as the model zoo.

## Modification
- [x] Rewrite the `YOLOXHead.get_bboxes`, rewrite `backbones.csp_darknet.Focus.forward`.
- [x] Add the docs of rewriters.
- [x] Add the unittests.
- [x] Static 416x416 shape evaluate and android performance test and update the docs.
## BC-breaking (Optional)

No.

## Use cases

```
python deploy.py ../configs/mmdet/detection/single-stage_ncnn_static-416x416.py ../../../data_example/mmdet/configs/yolox/yolox_tiny_8x8_300e_coco.py ../../../data_example/mmdet/checkpoints/yolox_tiny_8x8_300e_coco_20211124_171234-b4047906.pth ../../../data_example/mmdet/demo_416x416.jpg --work-dir ./
```

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by RunningLeon** [comment]
@hanrui1sensetime Hi, have you test the speed performance?

**Comment by HenryHan960829** [comment]
> @hanrui1sensetime Hi, have you test the speed performance?

In our PC cpu, it is 0.7s/task.
Android device has not tested yet.

**Comment by hanrui1sensetime** [comment]
In our PC cpu, it is 0.7s/task.
Android device has not tested yet.

**Comment by hanrui1sensetime** [comment]
> @hanrui1sensetime Hi, have you test the speed performance?

For android Adreno-660, it is 124.93ms and 8.00FPS. For android SnapDragon888, it is 132.69ms, and 7.54FPS.

File: mmdeploy/codebase/mmdet/models/backbones.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
+import torch
```
[the remark this answers is not in the document]
Comment (hanrui1sensetime) [reply mmdeploy/codebase/mmdet/models/backbones.py:1]: backbones rewriting is only used for this case, like neck rewriting is used for ssd. So, only a backbone.py file is enough.

File: mmdeploy/codebase/mmdet/models/dense_heads/yolox_head.py [hunk 2]
```diff
@@ -91,3 +91,124 @@ def yolox_head__get_bboxes(ctx,
...
+        rescale (bool): If True, return boxes in original image space.
+            Default False.
+        with_nms (bool): If True, do nms before return boxes.
+            Default True.
+
+    Returns:
+        output__ncnn (Tensor): outputs, shape is [N, num_det, 6].
+    """
+    from mmdeploy.codebase.mmdet.core.ops import ncnn_detection_output_forward
+    if isinstance(img_metas[0]['img_shape'][0], int):
```
[the remark this answers is not in the document]
Comment (hanrui1sensetime) [reply mmdeploy/codebase/mmdet/models/dense_heads/yolox_head.py:142]: > https://github.com/open-mmlab/mmdeploy/blob/f203306cf64266d4bee5cb43ea81a37bc506c6dc/mmdeploy/codebase/mmdet/models/detectors/base.py#L23

In my yolox_head.py code, things are different. the get_bboxes function cannot get the img data. So this way may not work.

File: mmdeploy/codebase/mmdet/models/dense_heads/yolox_head.py [hunk 3]
```diff
@@ -91,3 +91,124 @@ def yolox_head__get_bboxes(ctx,
...
+    pre_top_k = post_params.pre_top_k
+    keep_top_k = cfg.get('max_per_img', post_params.keep_top_k)
+
+    vars = torch.tensor([1, 1, 1, 1], dtype=torch.float32)
+    output__ncnn = ncnn_detection_output_forward(
+        batch_mlvl_bboxes, batch_mlvl_scores, batch_mlvl_priors,
+        score_threshold, iou_threshold, pre_top_k, keep_top_k,
+        self.num_classes + 1,
+        vars.cpu().detach().numpy())
+    return output__ncnn
```
[the remark this answers is not in the document]
Comment (hanrui1sensetime) [reply mmdeploy/codebase/mmdet/models/dense_heads/yolox_head.py:215]: I and @SingleZombie have tested this model could deploy. May be some environment error?

File: mmdeploy/codebase/mmdet/models/backbones.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        x (Tensor): The input tensor with shape (N, C, H, W).
+
+    Returns:
+        x (Tensor): The calculated tensor with shape (N, 4*C, H//2, W//2).
+    """
+    batch_size, c, h, w = x.shape
+    assert h % 2 == 0 and w % 2 == 0, f'focus for yolox needs even feature\
+        height and width, got {(h, w)}.'
+
+    x_col = x.reshape(batch_size, c, -1, 2)
```
Comment (grimoire) [inline mmdeploy/codebase/mmdet/models/backbones.py:26]: Can we use `PixelShuffle` or `ShuffleChannel` in ncnn?

File: mmdeploy/codebase/mmdet/models/dense_heads/yolox_head.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (grimoire) [reply mmdeploy/codebase/mmdet/models/dense_heads/yolox_head.py:142]: Could use `int(img_metas[0]['img_shape'][0])` and remove the branch.

File: mmdeploy/codebase/mmdet/models/backbones.py (hunk 4 above)
Comment (hanrui1sensetime) [reply mmdeploy/codebase/mmdet/models/backbones.py:26]: > Can we use `PixelShuffle` or `ShuffleChannel` in ncnn?

Seems that ncnn `PixelShuffle` only used for upsample case but not this downsample case.
`ShuffleChannel` can use in ncnn.

**Review by grimoire — APPROVED** [review]

File: mmdeploy/codebase/mmdet/models/dense_heads/yolox_head.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (RunningLeon) [reply mmdeploy/codebase/mmdet/models/dense_heads/yolox_head.py:211]:
```
2022-01-10:11:38:23,root INFO     [deploy.py:72] torch2onnx start.
2022-01-10:11:38:27,root ERROR    [deploy.py:67] YOLOX: __init__() got an unexpected keyword argument 'input_size'
Traceback (most recent call last):
  File "/home/PJLAB/maningsheng/projects/openmmlab/mmcv-pt1.8/mmcv/utils/registry.py", line 52, in build_from_cfg
    return obj_cls(**args)
TypeError: __init__() got an unexpected keyword argument 'input_size'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/PJLAB/maningsheng/projects/openmmlab/mmdeploy/tools/deploy.py", line 62, in target_wrapper
    result = target(*args, **kwargs)
  File "/home/PJLAB/maningsheng/projects/openmmlab/mmdeploy/mmdeploy/apis/pytorch2onnx.py", line 87, in torch2onnx
    torch_model = task_processor.init_pytorch_model(model_checkpoint)
  File "/home/PJLAB/maningsheng/projects/openmmlab/mmdeploy/mmdeploy/codebase/mmdet/deploy/object_detection.py", line 95, in init_pytorch_model
    cfg_options)
  File "/home/PJLAB/maningsheng/anaconda3/envs/yolov3/lib/python3.7/site-packages/mmdet/apis/inference.py", line 40, in init_detector
    model = build_detector(config.model, test_cfg=config.get('test_cfg'))
  File "/home/PJLAB/maningsheng/anaconda3/envs/yolov3/lib/python3.7/site-packages/mmdet/models/builder.py", line 59, in build_detector
    cfg, default_args=dict(train_cfg=train_cfg, test_cfg=test_cfg))
  File "/home/PJLAB/maningsheng/projects/openmmlab/mmcv-pt1.8/mmcv/utils/registry.py", line 212, in build
    return self.build_func(*args, **kwargs, registry=self)
  File "/home/PJLAB/maningsheng/projects/openmmlab/mmcv-pt1.8/mmcv/cnn/builder.py", line 27, in build_model_from_cfg
    return build_from_cfg(cfg, registry, default_args)
  File "/home/PJLAB/maningsheng/projects/openmmlab/mmcv-pt1.8/mmcv/utils/registry.py", line 55, in build_from_cfg
    raise type(e)(f'{obj_cls.__name__}: {e}')
TypeError: YOLOX: __init__() got an unexpected keyword argument 'input_size'
2022-01-10:11:38:27,root ERROR    [deploy.py:83] torch2onnx failed.

Process finished with exit code 0
```

**Comment by RunningLeon** [comment]
@hanrui1sensetime May fix conflicts. BTW, does it work now for yolox ncnn?

File: mmdeploy/codebase/mmdet/models/backbones.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    assert h % 2 == 0 and w % 2 == 0, f'focus for yolox needs even feature\
+        height and width, got {(h, w)}.'
+
+    x = x.reshape(batch_size, c * h, 1, w)
+    _b, _c, _h, _w = x.shape
+    g = _c // 2
+    x = x.view(_b, g, 2, _h, _w)
+    x = torch.transpose(x, 1, 2).contiguous()
+    # fuse to ncnn's shufflechannel
+    x = x.view(_b, -1, _h, _w)
```
Comment (grimoire) [inline mmdeploy/codebase/mmdet/models/backbones.py:32]: why don't we just reshape to `(_b, c * h * w, 1, 1)` ?

**Comment by hanrui1sensetime** [comment]
> @hanrui1sensetime May fix conflicts. BTW, does it work now for yolox ncnn?

Now, the model supports dynamic shape and static shape (e.g. 416x416). I have dropped out the pad_to_square in static shape.

File: mmdeploy/backend/ncnn/wrapper.py [hunk 6]
```diff
@@ -112,11 +112,15 @@ def forward(self, inputs: Dict[str,
                     mat = None
```
Comment (RunningLeon) [inline mmdeploy/backend/ncnn/wrapper.py:112]: This line is unused. could remove.

File: mmdeploy/codebase/mmdet/deploy/object_detection.py [hunk 7]
```diff
@@ -114,7 +114,20 @@ def create_input(self,
...
+        dynamic_flag = is_dynamic_shape(self.deploy_cfg)
         cfg = process_model_config(self.model_cfg, imgs, input_shape)
+        # Drop pad_to_square when static shape. Because static shape should
+        # ensure the shape before input image.
+        if not dynamic_flag:
+            transform = cfg.data.test.pipeline[1]
+            if 'transforms' in transform:
+                transform_list = transform['transforms']
+                pos = 0
+                for step in transform_list:
```
Comment (RunningLeon) [inline mmdeploy/codebase/mmdet/deploy/object_detection.py:126]: may find in for loop and remove outside for loop.

**Comment by RunningLeon** [comment]
### NCNN results: bbox_mAP: 32.0
```
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.320
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=1000 ] = 0.494
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=1000 ] = 0.340
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=1000 ] = 0.131
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=1000 ] = 0.351
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=1000 ] = 0.478
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.462
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=300 ] = 0.462
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=1000 ] = 0.462
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=1000 ] = 0.209
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=1000 ] = 0.515
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=1000 ] = 0.666

OrderedDict([('bbox_mAP', 0.32), ('bbox_mAP_50', 0.494), ('bbox_mAP_75', 0.34), ('bbox_mAP_s', 0.131), ('bbox_mAP_m', 0.351), ('bbox_mAP_l', 0.478), ('bbox_mAP_copypaste', '0.320 0.494 0.340 0.131 0.351 0.478')])
```

### pytorch results:

| Backbone  | size   | Mem (GB) |   box AP | Config | Download |
|:---------:|:-------:|:-------:|:-------:|:--------:|:------:|
| YOLOX-tiny | 416 |   3.5      |   32.0  | [config](https://github.com/open-mmlab/mmdetection/tree/master/configs/yolox/yolox_tiny_8x8_300e_coco.py)       |[model](https://download.openmmlab.com/mmdetection/v2.0/yolox/yolox_tiny_8x8_300e_coco/yolox_tiny_8x8_300e_coco_20211124_171234-b4047906.pth) &#124; [log](https://download.openmmlab.com/mmdetection/v2.0/yolox/yolox_tiny_8x8_300e_coco/yolox_tiny_8x8_300e_coco_20211124_171234.log.json) |

File: docs/en/supported_models.md [hunk 8]
```diff
@@ -7,7 +7,7 @@ The table below lists the models that are guaranteed to be exportable to other b
 | RetinaNet                 | MMDetection      |      Y      |    Y     |   Y   |   Y   |    Y     |       [config](https://github.com/open-mmlab/mmdetection/tree/master/configs/retinanet)        |
 | Faster R-CNN              | MMDetection      |      Y      |    Y     |   Y   |   Y   |    Y     |      [config](https://github.com/open-mmlab/mmdetection/tree/master/configs/faster_rcnn)       |
 | YOLOv3                    | MMDetection      |      Y      |    Y     |   Y   |   N   |    Y     |          [config](https://github.com/open-mmlab/mmdetection/tree/master/configs/yolo)          |
-| YOLOX                     | MMDetection      |      Y      |    Y     |   N   |   N   |    Y     |         [config](https://github.com/open-mmlab/mmdetection/tree/master/configs/yolox)          |
+| YOLOX                     | MMDetection      |      Y      |    Y     |   Y   |   N   |    Y     |         [config](https://github.com/open-mmlab/mmdetection/tree/master/configs/yolox)          |
```
Comment (RunningLeon) [inline docs/en/supported_models.md:10]: may need to update this table as well:
https://github.com/open-mmlab/mmdeploy/blob/master/docs/en/codebases/mmdet.md#list-of-mmdetection-models-supported-by-mmdeploy

File: mmdeploy/codebase/mmdet/deploy/object_detection.py [hunk 9]
```diff
@@ -114,7 +114,23 @@ def create_input(self,
...
         cfg = process_model_config(self.model_cfg, imgs, input_shape)
+        # Drop pad_to_square when static shape. Because static shape should
+        # ensure the shape before input image.
+        if not dynamic_flag:
+            transform = cfg.data.test.pipeline[1]
+            if 'transforms' in transform:
+                transform_list = transform['transforms']
+                ind = []
+                pos = 0
+                for step in transform_list:
```
Comment (RunningLeon) [inline mmdeploy/codebase/mmdet/deploy/object_detection.py:127]: for i, step in enumerate(transform_list):
    if step['type'] == 'Pad' and 'pad_to_square' in step \
                       and step['pad_to_square']:
        transform_list.pop(i)
        break

File: mmdeploy/codebase/mmdet/models/dense_heads/yolox_head.py [hunk 10]
```diff
@@ -91,3 +91,121 @@ def yolox_head__get_bboxes(ctx,
...
+        rescale (bool): If True, return boxes in original image space.
+            Default False.
+        with_nms (bool): If True, do nms before return boxes.
+            Default True.
+
+    Returns:
+        output__ncnn (Tensor): outputs, shape is [N, num_det, 6].
+    """
+    from mmdeploy.codebase.mmdet.core.ops import ncnn_detection_output_forward
+    img_height = int(img_metas[0]['img_shape'][0])
```
Comment (RunningLeon) [inline mmdeploy/codebase/mmdet/models/dense_heads/yolox_head.py:142]: yolox ncnn does not support dynamic shape. may add warning if input config is dynamic .

File: mmdeploy/codebase/mmdet/models/dense_heads/yolox_head.py [hunk 11]
```diff
@@ -139,6 +139,13 @@ def yolox_head__get_bboxes__ncnn(ctx,
         output__ncnn (Tensor): outputs, shape is [N, num_det, 6].
     """
     from mmdeploy.codebase.mmdet.core.ops import ncnn_detection_output_forward
+    from mmdeploy.utils.config_utils import is_dynamic_shape
+    from mmdeploy.utils import get_root_logger
+    dynamic_flag = is_dynamic_shape(ctx.cfg)
+    if dynamic_flag:
+        logger = get_root_logger()
+        logger.warning('YOLOX model fix the size of (416, 416), \
```
Comment (RunningLeon) [inline mmdeploy/codebase/mmdet/models/dense_heads/yolox_head.py:147]: logger.warning('YOLOX does not support dynamic shape with ncnn.')

File: docs/en/supported_models.md (hunk 8 above)
Comment (RunningLeon) [inline docs/en/supported_models.md:10]: may add one more **note** at the end of doc that `YOLOX with ncnn only supports static shape.`

File: tests/test_codebase/test_mmdet/test_mmdet_models.py [hunk 12]
```diff
@@ -112,6 +112,15 @@ def get_fcos_head_model():
     return model
 
 
+def get_focus_backbone_model():
+    """FCOS Head Config."""
```
Comment (RunningLeon) [inline tests/test_codebase/test_mmdet/test_mmdet_models.py:116]: wrong docstring. need to update.

**Review by RunningLeon — APPROVED** [review]
LGTM

**Pull request merged**